File: ThreeEngine/models/Animation.cpp
#include <algorithm>
#include "Animation.h"

template<class T>
inline T linear_interpolate(T y0, T y1, float t)
{
	return (1.0f - t)* y0 + t * y1;
}

template<class T>
inline T cubic_interpolate(T y0, T y1, T s0, T s1, float t, float t2, float t3, float td)
{
	return (2.0f * t3 - 3.0f * t2 + 1) * y0 + (-2.0f * t3 + 3.0f * t2) * y1	+ td * ((t3 - 2.0f * t2 + t) * s0 + (t3 - t2) * s1);
}

void MorphTrack::get_frame(float x, MorphFrame& frame)
{
	frame.name = name;
	frame.weights.resize(num_targets);
	auto iter = std::upper_bound(times.begin(), times.end(), x);	

	if (iter == times.begin())
	{
		if (interpolation == Interpolation::STEP || interpolation == Interpolation::LINEAR)
		{
			memcpy(frame.weights.data(), values.data(), sizeof(float) * num_targets);
		}
		else if (interpolation == Interpolation::CUBICSPLINE)
		{
			memcpy(frame.weights.data(), values.data() + num_targets, sizeof(float) * num_targets);
		}
	}
	else if (iter == times.end())
	{
		if (interpolation == Interpolation::STEP || interpolation == Interpolation::LINEAR)
		{
			memcpy(frame.weights.data(), values.data() + num_targets*(times.size()-1), sizeof(float) * num_targets);
		}
		else if (interpolation == Interpolation::CUBICSPLINE)
		{
			memcpy(frame.weights.data(), values.data() + num_targets + num_targets * (times.size() - 1)*3, sizeof(float) * num_targets);
		}
	}
	else
	{
		float x0 = *(iter - 1);
		float x1 = *iter;
		if (interpolation == Interpolation::STEP)
		{
			memcpy(frame.weights.data(), values.data() + num_targets*(iter- times.begin()-1), sizeof(float) * num_targets);
		}
		else if (interpolation == Interpolation::LINEAR)
		{
			const float* p0 = values.data() + num_targets * (iter - times.begin() - 1);
			const float* p1 = values.data() + num_targets * (iter - times.begin());
			float t = (x - x0) / (x1 - x0);
			for (int i = 0; i < num_targets; i++)
			{
				frame.weights[i] = linear_interpolate(p0[i], p1[i], t);
			}
		}
		else if (interpolation == Interpolation::CUBICSPLINE)
		{
			const float* p0 = values.data() + num_targets * (iter - times.begin() - 1) * 3;
			const float* p1 = values.data() + num_targets * (iter - times.begin()) * 3;
			float t = (x - x0) / (x1 - x0);
			float t2 = t * t;
			float t3 = t2 * t;
			float td = x1 - x0;
			for (int i = 0; i < num_targets; i++)
			{
				float y0 = p0[num_targets + i];
				float s0 = p0[num_targets*2 + i];
				float s1 = p1[i];
				float y1 = p1[num_targets + i];

				frame.weights[i] = cubic_interpolate(y0, y1, s0, s1, t, t2, t3, td);
			}
		}
	}
}

void TranslationTrack::get_frame(float x, TranslationFrame& frame)
{
	frame.name = name;	

	auto iter = std::upper_bound(times.begin(), times.end(), x);
	if (iter == times.begin())
	{
		if (interpolation == Interpolation::STEP || interpolation == Interpolation::LINEAR)
		{
			frame.translation = values[0];
		}
		else if (interpolation == Interpolation::CUBICSPLINE)
		{
			frame.translation = values[1];
		}
	}
	else if (iter == times.end())
	{
		if (interpolation == Interpolation::STEP || interpolation == Interpolation::LINEAR)
		{
			frame.translation = values[times.size() - 1];
		}
		else if (interpolation == Interpolation::CUBICSPLINE)
		{
			frame.translation = values[1 + (times.size() - 1) * 3];			
		}	
	}
	else
	{
		float x0 = *(iter - 1);
		float x1 = *iter;
		if (interpolation == Interpolation::STEP)
		{
			frame.translation = values[iter - times.begin() - 1];
		}
		else if (interpolation == Interpolation::LINEAR)
		{
			const glm::vec3& y0 = values[iter - times.begin() - 1];
			const glm::vec3& y1 = values[iter - times.begin()];
			float t = (x - x0) / (x1 - x0);
			frame.translation = linear_interpolate(y0, y1, t);
		}
		else if (interpolation == Interpolation::CUBICSPLINE)
		{
			float t = (x - x0) / (x1 - x0);
			float t2 = t * t;
			float t3 = t2 * t;
			float td = x1 - x0;

			const glm::vec3* p0 = values.data() + (iter - times.begin() - 1) * 3;
			const glm::vec3* p1 = values.data() + (iter - times.begin()) * 3;
			const glm::vec3& y0 = p0[1];
			const glm::vec3& s0 = p0[2];
			const glm::vec3& s1 = p1[0];
			const glm::vec3& y1 = p1[1];

			frame.translation = cubic_interpolate(y0, y1, s0, s1, t, t2, t3, td);
		}
	}
}

void RotationTrack::get_frame(float x, RotationFrame& frame)
{
	frame.name = name;

	auto iter = std::upper_bound(times.begin(), times.end(), x);
	if (iter == times.begin())
	{
		if (interpolation == Interpolation::STEP || interpolation == Interpolation::LINEAR)
		{
			frame.rotation = values[0];
		}
		else if (interpolation == Interpolation::CUBICSPLINE)
		{
			frame.rotation = values[1];
		}
	}
	else if (iter == times.end())
	{
		if (interpolation == Interpolation::STEP || interpolation == Interpolation::LINEAR)
		{
			frame.rotation = values[times.size() - 1];
		}
		else if (interpolation == Interpolation::CUBICSPLINE)
		{
			frame.rotation = values[1 + (times.size() - 1) * 3];
		}		
	}
	else
	{
		float x0 = *(iter - 1);
		float x1 = *iter;
		if (interpolation == Interpolation::STEP)
		{
			frame.rotation = values[iter - times.begin() - 1];
		}
		else if (interpolation == Interpolation::LINEAR)
		{
			const glm::quat& y0 = values[iter - times.begin() - 1];
			const glm::quat& y1 = values[iter - times.begin()];
			float t = (x - x0) / (x1 - x0);
			frame.rotation = glm::slerp(y0, y1, t);
		}
		else if (interpolation == Interpolation::CUBICSPLINE)
		{
			float t = (x - x0) / (x1 - x0);		
			const glm::quat* p0 = values.data() + (iter - times.begin() - 1) * 3;
			const glm::quat* p1 = values.data() + (iter - times.begin()) * 3;
			const glm::quat& y0 = p0[1];
			const glm::quat& y1 = p1[1];
			frame.rotation = glm::slerp(y0, y1, t);
		}
	}
}


void ScaleTrack::get_frame(float x, ScaleFrame& frame)
{
	frame.name = name;

	auto iter = std::upper_bound(times.begin(), times.end(), x);
	if (iter == times.begin())
	{
		if (interpolation == Interpolation::STEP || interpolation == Interpolation::LINEAR)
		{
			frame.scale = values[0];
		}
		else if (interpolation == Interpolation::CUBICSPLINE)
		{
			frame.scale = values[1];
		}
	}
	else if (iter == times.end())
	{
		if (interpolation == Interpolation::STEP || interpolation == Interpolation::LINEAR)
		{
			frame.scale = values[times.size() - 1];
		}
		else if (interpolation == Interpolation::CUBICSPLINE)
		{
			frame.scale = values[1 + (times.size() - 1) * 3];
		}		
	}
	else
	{
		float x0 = *(iter - 1);
		float x1 = *iter;
		if (interpolation == Interpolation::STEP)
		{
			frame.scale = values[iter - times.begin() - 1];
		}
		else if (interpolation == Interpolation::LINEAR)
		{
			const glm::vec3& y0 = values[iter - times.begin() - 1];
			const glm::vec3& y1 = values[iter - times.begin()];
			float t = (x - x0) / (x1 - x0);
			frame.scale = linear_interpolate(y0, y1, t);
		}
		else if (interpolation == Interpolation::CUBICSPLINE)
		{
			float t = (x - x0) / (x1 - x0);
			float t2 = t * t;
			float t3 = t2 * t;
			float td = x1 - x0;

			const glm::vec3* p0 = values.data() + (iter - times.begin() - 1) * 3;
			const glm::vec3* p1 = values.data() + (iter - times.begin()) * 3;
			const glm::vec3& y0 = p0[1];
			const glm::vec3& s0 = p0[2];
			const glm::vec3& s1 = p1[0];
			const glm::vec3& y1 = p1[1];

			frame.scale = cubic_interpolate(y0, y1, s0, s1, t, t2, t3, td);
		}
	}
}


void AnimationClip::get_frame(float x, AnimationFrame& frame)
{
	// morphs
	size_t num_morphs = morphs.size();
	frame.morphs.resize(num_morphs);	
	for (size_t i = 0; i < num_morphs; i++)
	{
		morphs[i].get_frame(x, frame.morphs[i]);
	}

	// translations
	size_t num_trans = translations.size();
	frame.translations.resize(num_trans);
	for (size_t i = 0; i < num_trans; i++)
	{
		translations[i].get_frame(x, frame.translations[i]);
	}

	// rotations
	size_t num_rots = rotations.size();
	frame.rotations.resize(num_rots);
	for (size_t i = 0; i < num_rots; i++)
	{
		rotations[i].get_frame(x, frame.rotations[i]);
	}

	// scales
	size_t num_scales = scales.size();
	frame.scales.resize(num_scales);
	for (size_t i = 0; i < num_scales; i++)
	{
		scales[i].get_frame(x, frame.scales[i]);
	}
}

File: webgpu/client/engine/models/Animation.js
import { Vector3 } from "../math/Vector3.js"
import { Quaternion } from "../math/Quaternion.js";

export class MorphFrame
{
    constructor()
    {
        this.name = "";
        this.weights = [];
    }
}

export class TranslationFrame
{
    constructor()
    {
        this.name = "";
        this.translation = new Vector3(0,0,0);
    }
}

export class RotationFrame
{
    constructor()
    {
        this.name = "";
        this.rotation = new Quaternion(0,0,0,1);
    }
}

export class ScaleFrame
{
    constructor()
    {
        this.name = "";
        this.scale = new Vector3(1,1,1);
    }
}

export class AnimationFrame
{
    constructor()
    {
        this.morphs = [];
        this.translations = [];
        this.rotations = [];
        this.scales = [];
    }
}

function linear_interpolate(y0, y1, t)
{
    return (1.0 - t)*y0 + t*y1;
}

function cubic_interpolate(y0, y1, s0, s1, t, t2, t3, td)
{
    return (2.0 * t3 - 3.0 * t2 + 1.0) * y0 + (-2.0 * t3 + 3.0 * t2) * y1	+ td * ((t3 - 2.0 * t2 + t) * s0 + (t3 - t2) * s1);
}

function upperbound(arr, target)
{
    var l = 0;
    var r = arr.length;
    while(l<r)
    {
        var mid = Math.floor((l+r)/2);                    
        if(arr[mid] <= target)
        {
            l = mid+1;
        }
        else
        {
            r = mid;
        }
    }
    return l;
}

export class MorphTrack
{
    constructor()
    {
        this.name = "";
        this.targets = 0;
        this.interpolation = "LINEAR";
        this.times = new Float32Array();
        this.values = new Float32Array();
    }

    get_frame(x)
    {
        let frame = new MorphFrame();
        frame.name = this.name;
        frame.weights = new Array(this.targets);

        let idx = upperbound(this.times, x);

        if (idx==0)
        {
            if (this.interpolation == "STEP" || this.interpolation == "LINEAR")
            {
                for (let i=0; i<this.targets; i++)
                {
                    frame.weights[i] = this.values[i];
                }
            }
            else if (this.interpolation == "CUBICSPLINE")
            {
                let offset = this.targets;
                for (let i=0; i<this.targets; i++)
                {
                    frame.weights[i] = this.values[offset + i];
                }
            }
        }
        else if (idx == this.times.length)
        {
            if (this.interpolation == "STEP" || this.interpolation == "LINEAR")
            {
                let offset = this.targets *(this.times.length-1);
                for (let i=0; i<this.targets; i++)
                {
                    frame.weights[i] = this.values[offset + i];
                }
            }
            else if (this.interpolation == "CUBICSPLINE")
            {
                let offset = this.targets *(this.times.length-1)*3 + this.targets;
                for (let i=0; i<this.targets; i++)
                {
                    frame.weights[i] = this.values[offset + i];
                }
            }
        }
        else
        {
            let x0 = this.times[idx-1];
            let x1 = this.times[idx];
            if (this.interpolation == "STEP")
            {
                let offset = this.targets * (idx-1);
                for (let i=0; i<this.targets; i++)
                {
                    frame.weights[i] = this.values[offset + i];
                }
            }
            else if (this.interpolation == "LINEAR")
            {
                let offset0 = this.targets * (idx-1);
                let offset1 = this.targets * idx;
                let t = (x-x0)/(x1-x0);
                for (let i=0; i<this.targets; i++)
                {
                    let y0 = this.values[offset0 + i];
                    let y1 = this.values[offset1 + i];
                    frame.weights[i] = linear_interpolate(y0, y1, t);
                }
            }
            else if(this.interpolation == "CUBICSPLINE")
            {
                let offset0 =  this.targets * (idx-1) * 3;
                let offset1 =  this.targets * idx * 3;
                let t = (x-x0)/(x1-x0);
                let t2 = t*t;
                let t3 = t2*t;
                let td = x1 - x0;
                for (let i=0; i<this.targets; i++)
                {
                    let y0 = this.values[offset0 + this.targets + i];
                    let s0 = this.values[offset0 + this.targets*2 + i];                    
                    let s1 = this.values[offset1 + i];
                    let y1 = this.values[offset1 + this.targets + i];
                    frame.weights[i] = cubic_interpolate(y0, y1, s0, s1, t, t2, t3, td);
                }
            }
        }

        return frame;
    }
}

export class TranslationTrack
{
    constructor()
    {
        this.name = "";        
        this.interpolation = "LINEAR";
        this.times = new Float32Array();
        this.values = new Float32Array();
    }

    get_frame(x)
    {
        let frame = new TranslationFrame();
        frame.name = this.name;

        let idx = upperbound(this.times, x);
        if (idx==0)
        {
            if (this.interpolation == "STEP" || this.interpolation == "LINEAR")
            {
                frame.translation.set(this.values[0], this.values[1], this.values[2]);
            }
            else if (this.interpolation == "CUBICSPLINE")
            {
                frame.translation.set(this.values[3], this.values[4], this.values[5]);
            }
        }
        else if (idx == this.times.length)
        {
            if (this.interpolation == "STEP" || this.interpolation == "LINEAR")
            {
                let offset = 3 *(this.times.length-1);
                frame.translation.set(this.values[offset], this.values[offset+1], this.values[offset+2]);
            }
            else if (this.interpolation == "CUBICSPLINE")
            {
                let offset = 3 *(this.times.length-1)*3 + 3;
                frame.translation.set(this.values[offset], this.values[offset+1], this.values[offset+2]);
            }
        }
        else
        {
            let x0 = this.times[idx-1];
            let x1 = this.times[idx];
            if (this.interpolation == "STEP")
            {
                let offset = 3 * (idx-1);
                frame.translation.set(this.values[offset], this.values[offset+1], this.values[offset+2]);
            }
            else if (this.interpolation == "LINEAR")
            {
                let offset0 = 3 * (idx-1);
                let offset1 = 3 * idx;
                let t = (x-x0)/(x1-x0);                
                {
                    let y0 = this.values[offset0];
                    let y1 = this.values[offset1];
                    frame.translation.x = linear_interpolate(y0, y1, t);
                }
                {
                    let y0 = this.values[offset0+1];
                    let y1 = this.values[offset1+1];
                    frame.translation.y = linear_interpolate(y0, y1, t);
                }
                {
                    let y0 = this.values[offset0+2];
                    let y1 = this.values[offset1+2];
                    frame.translation.z = linear_interpolate(y0, y1, t);
                }
            }
            else if(this.interpolation == "CUBICSPLINE")
            {
                let offset0 =  3 * (idx-1) * 3;
                let offset1 =  3 * idx * 3;
                let t = (x-x0)/(x1-x0);
                let t2 = t*t;
                let t3 = t2*t;
                let td = x1 - x0;
                {
                    let y0 = this.values[offset0 + 3];
                    let s0 = this.values[offset0 + 3*2];
                    let s1 = this.values[offset1];
                    let y1 = this.values[offset1 + 3];
                    frame.translation.x = cubic_interpolate(y0, y1, s0, s1, t, t2, t3, td);
                }

                {
                    let y0 = this.values[offset0 + 3 + 1];
                    let s0 = this.values[offset0 + 3*2 + 1];
                    let s1 = this.values[offset1 + 1];
                    let y1 = this.values[offset1 + 3 + 1];
                    frame.translation.y = cubic_interpolate(y0, y1, s0, s1, t, t2, t3, td);
                }

                {
                    let y0 = this.values[offset0 + 3 + 2];
                    let s0 = this.values[offset0 + 3*2 + 2];
                    let s1 = this.values[offset1 + 2];
                    let y1 = this.values[offset1 + 3 + 2];
                    frame.translation.z = cubic_interpolate(y0, y1, s0, s1, t, t2, t3, td);
                }                
            }
        }
        return frame;
    }
}

export class RotationTrack
{
    constructor()
    {
        this.name = "";        
        this.interpolation = "LINEAR";
        this.times = new Float32Array();
        this.values = new Float32Array();
    }

    get_frame(x)
    {
        let frame = new RotationFrame();
        frame.name = this.name;

        let idx = upperbound(this.times, x);
        if (idx==0)
        {
            if (this.interpolation == "STEP" || this.interpolation == "LINEAR")
            {
                frame.rotation.set(this.values[0], this.values[1], this.values[2], this.values[3]);
            }
            else if (this.interpolation == "CUBICSPLINE")
            {
                frame.rotation.set(this.values[4], this.values[5], this.values[6], this.values[7]);
            }
        }
        else if (idx == this.times.length)
        {
            if (this.interpolation == "STEP" || this.interpolation == "LINEAR")
            {
                let offset = 4 *(this.times.length-1);
                frame.rotation.set(this.values[offset], this.values[offset+1], this.values[offset+2], this.values[offset+3]);
            }
            else if (this.interpolation == "CUBICSPLINE")
            {
                let offset = 4 *(this.times.length-1)*3 + 4;
                frame.rotation.set(this.values[offset], this.values[offset+1], this.values[offset+2], this.values[offset+3]);
            }
        }
        else
        {
            let x0 = this.times[idx-1];
            let x1 = this.times[idx];
            if (this.interpolation == "STEP")
            {
                let offset = 4 * (idx-1);
                frame.rotation.set(this.values[offset], this.values[offset+1], this.values[offset+2], this.values[offset+3]);
            }
            else if (this.interpolation == "LINEAR")
            {
                let offset0 = 4 * (idx-1);
                let offset1 = 4 * idx;
                let t = (x-x0)/(x1-x0);

                let y0 = new Quaternion(
                    this.values[offset0], 
                    this.values[offset0 + 1],
                    this.values[offset0 + 2],
                    this.values[offset0 + 3]);

                let y1 = new Quaternion(
                    this.values[offset1], 
                    this.values[offset1 + 1],
                    this.values[offset1 + 2],
                    this.values[offset1 + 3]);

                frame.rotation.slerpQuaternions(y0, y1, t);
            }
            else if(this.interpolation == "CUBICSPLINE")
            {
                let offset0 =  4 * (idx-1) * 3;
                let offset1 =  4 * idx * 3;
                let t = (x-x0)/(x1-x0);

                let y0 = new Quaternion(
                    this.values[offset0 + 4], 
                    this.values[offset0 + 4 + 1],
                    this.values[offset0 + 4 + 2],
                    this.values[offset0 + 4 + 3]);

                let y1 = new Quaternion(
                    this.values[offset1 + 4], 
                    this.values[offset1 + 4 + 1],
                    this.values[offset1 + 4 + 2],
                    this.values[offset1 + 4 + 3]);
                
                frame.rotation.slerpQuaternions(y0, y1, t);
            }
        }
        return frame;
    }
}

export class ScaleTrack
{
    constructor()
    {
        this.name = "";        
        this.interpolation = "LINEAR";
        this.times = new Float32Array();
        this.values = new Float32Array();
    }
    
    get_frame(x)
    {
        let frame = new ScaleFrame();
        frame.name = this.name;

        let idx = upperbound(this.times, x);
        if (idx==0)
        {
            if (this.interpolation == "STEP" || this.interpolation == "LINEAR")
            {
                frame.scale.set(this.values[0], this.values[1], this.values[2]);
            }
            else if (this.interpolation == "CUBICSPLINE")
            {
                frame.scale.set(this.values[3], this.values[4], this.values[5]);
            }
        }
        else if (idx == this.times.length)
        {
            if (this.interpolation == "STEP" || this.interpolation == "LINEAR")
            {
                let offset = 3 *(this.times.length-1);
                frame.scale.set(this.values[offset], this.values[offset+1], this.values[offset+2]);
            }
            else if (this.interpolation == "CUBICSPLINE")
            {
                let offset = 3 *(this.times.length-1)*3 + 3;
                frame.scale.set(this.values[offset], this.values[offset+1], this.values[offset+2]);
            }
        }
        else
        {
            let x0 = this.times[idx-1];
            let x1 = this.times[idx];
            if (this.interpolation == "STEP")
            {
                let offset = 3 * (idx-1);
                frame.scale.set(this.values[offset], this.values[offset+1], this.values[offset+2]);
            }
            else if (this.interpolation == "LINEAR")
            {
                let offset0 = 3 * (idx-1);
                let offset1 = 3 * idx;
                let t = (x-x0)/(x1-x0);                
                {
                    let y0 = this.values[offset0];
                    let y1 = this.values[offset1];
                    frame.scale.x = linear_interpolate(y0, y1, t);
                }
                {
                    let y0 = this.values[offset0+1];
                    let y1 = this.values[offset1+1];
                    frame.scale.y = linear_interpolate(y0, y1, t);
                }
                {
                    let y0 = this.values[offset0+2];
                    let y1 = this.values[offset1+2];
                    frame.scale.z = linear_interpolate(y0, y1, t);
                }
            }
            else if(this.interpolation == "CUBICSPLINE")
            {
                let offset0 =  3 * (idx-1) * 3;
                let offset1 =  3 * idx * 3;
                let t = (x-x0)/(x1-x0);
                let t2 = t*t;
                let t3 = t2*t;
                let td = x1 - x0;
                {
                    let y0 = this.values[offset0 + 3];
                    let s0 = this.values[offset0 + 3*2];
                    let s1 = this.values[offset1];
                    let y1 = this.values[offset1 + 3];
                    frame.scale.x = cubic_interpolate(y0, y1, s0, s1, t, t2, t3, td);
                }

                {
                    let y0 = this.values[offset0 + 3 + 1];
                    let s0 = this.values[offset0 + 3*2 + 1];
                    let s1 = this.values[offset1 + 1];
                    let y1 = this.values[offset1 + 3 + 1];
                    frame.scale.y = cubic_interpolate(y0, y1, s0, s1, t, t2, t3, td);
                }

                {
                    let y0 = this.values[offset0 + 3 + 2];
                    let s0 = this.values[offset0 + 3*2 + 2];
                    let s1 = this.values[offset1 + 2];
                    let y1 = this.values[offset1 + 3 + 2];
                    frame.scale.z = cubic_interpolate(y0, y1, s0, s1, t, t2, t3, td);
                }                
            }
        }
        return frame;
    }
}


export class AnimationClip
{
    constructor()
    {
        this.name = "";
        this.duration = 0;

        this.morphs = [];
        this.translations = [];
        this.rotations = [];
        this.scales = [];
    }

    get_frame(x)
    {
        let frame = new AnimationFrame();
        for (let morph of this.morphs)
        {
            frame.morphs.push(morph.get_frame(x));            
        }
        for (let trans of this.translations)
        {
            frame.translations.push(trans.get_frame(x));
        }
        for (let rot of this.rotations)
        {
            frame.rotations.push(rot.get_frame(x));
        }
        for (let scale of this.scales)
        {
            frame.scales.push(scale.get_frame(x));
        }

        return frame;
    }

}
